### PR TITLE
Add notification & poll support

### PR DESCRIPTION
**Pull request by zevweiss**
This patch series adds support for `poll()` and most of the currently-existing FUSE notification types (unsolicited messages sent by userspace to notify the kernel of various events).  It addresses issues #16, #17, and #160.

Not currently implemented are the `FUSE_EXPIRE_ONLY` flag for the `FUSE_NOTIFY_INVAL_ENTRY` operation (mostly just because it seems fairly niche and I had no particular need for it) and the `FUSE_NOTIFY_RETRIEVE` operation (since it involves bidirectional communication that would complicate the implementation).

~~I've also translated some of libfuse's example programs to exercise this functionality (see [this commit](https://github.com/zevweiss/fuser/commit/0d455a1cfa32d4febb99792036b985b2d0f11052) in my fork), though I haven't included them in this PR since they're directly derived from GPL code and are hence GPLed themselves, and I'm not sure what this project's feelings toward that would be -- if per-file exceptions for example programs would be acceptable I can add them as well.~~

Some examples exercising this functionality are also added, translated from the C code in libfuse's `example/` directory (and hence under the GPL, since they're directly derived from GPL code).

### CODE REVIEW

**Review by cberner — CHANGES_REQUESTED** [review]
Two minor comments.

Also, yes please go ahead and add an example to the `examples/` folder. Having a separate GPL license for that file(s) is fine, as long as it's only in `examples/`

Excited to have this implemented!

File: examples/simple.rs [hunk 1]
```diff
@@ -5,8 +5,8 @@ use clap::{crate_version, Arg, Command};
 use fuser::consts::FOPEN_DIRECT_IO;
 #[cfg(feature = "abi-7-26")]
 use fuser::consts::FUSE_HANDLE_KILLPRIV;
-#[cfg(feature = "abi-7-31")]
-use fuser::consts::FUSE_WRITE_KILL_PRIV;
+// #[cfg(feature = "abi-7-31")]
+// use fuser::consts::FUSE_WRITE_KILL_PRIV;
```
[the remark this answers is not in the document]
Comment (zevweiss) [reply examples/simple.rs:9]: The reasoning is mentioned in the commit message for commit 57cb8ef (if a bit confusingly, given the dual meanings of "use"):

> The use of FUSE_WRITE_KILL_PRIV is commented out, so comment out the
'use' for it as well, ...

Looking back in the git history, that commenting-out dates to commit 244547538cdc1ade2a268625600d44070d4b4049, when the commented-out use of it was added along with a non-commented `use` statement.

File: src/notify.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /// data in the given range)
+    #[cfg(feature = "abi-7-12")]
+    pub fn inval_inode(&self, ino: u64, offset: i64, len: i64) -> io::Result<()> {
+        let notif = Notification::new_inval_inode(ino, offset, len);
+        self.send_inval(notify_code::FUSE_NOTIFY_INVAL_INODE, &notif)
+    }
+
+    /// Update the kernel's cached copy of a given inode's data
+    #[cfg(feature = "abi-7-15")]
+    pub fn store(&self, ino: u64, offset: u64, data: &[u8]) -> io::Result<()> {
```
[the remark this answers is not in the document]
Comment (zevweiss) [reply src/notify.rs:49]: `i64` would be consistent with the `off_t` in libfuse and other offset parameters elsewhere (e.g.  in `fuse_notify_inval_inode_out`), but in this particular case the [kernel's uapi header](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/fuse.h?h=v6.6#n1027) and [the existing definition of `fuse_notify_store_out`](https://github.com/cberner/fuser/blob/8b390279a5da98df3f91b69aee8ef20de793a2fa/src/ll/fuse_abi.rs#L1080) both have it unsigned.  I'm not sure offhand about the reasoning for it (if there is any), but I stuck with what was there.  Thoughts?

[the remark this answers is not in the document]
Comment (cberner) [reply src/notify.rs:49]: Ah, let's leave it this way then

**Review by cberner — APPROVED** [review]

**Comment by zevweiss** [comment]
Any opinions on a preferred fix for the CI failure?  I'm not entirely certain of the purpose of the first `cargo build` that it's failing on there (just a toolchain/build.rs sanity check?); should we just add `examples/` to the first `ADD` line?

**Comment by cberner** [comment]
Ah, you can just change the Docker file so that it only builds the `simple` example binary. It uses that for running the tests

**Comment by cberner** [comment]
Or just deleting lines 20 and 21 would be fine too. That looks like a silly hack I added a while ago to make the Dockerfile build faster =P

**Comment by zevweiss** [comment]
Ah, okay -- I'll try just removing them.  (Unfortunately it looks like tweaking cargo command-line flags wouldn't work -- apparently it tries to find the files for any `[[example]]` sections unconditionally, and just treats anything missing as a hard parse failure no matter what flags are or aren't passed...)

**Comment by cberner** [comment]
Merged. Thanks!

**Comment by chris-laplante** [comment]
Thanks for this! Was just planning on implementing it myself :). Glad to see someone else did it.

**Pull request merged**